Avoid generating sort attributes for unsortable fields

OmarHawk opened the pull request:
e.g. MapstructExpression annotated direct and related fields

At the moment, JHipster generates the sortBy directive also on unsortable fields (transient fields, like mapstructExpression annotated fields) for relationships, which produces an internal server error, when users use the functionality. 
A similar problem is there for direct such fields - it still generates the sortBy without the icon. So JHipster generates something kind of inconsistent, though it does not lead to Internal Server errror when the user clicks it, as the directive has an if statement, preventing to sort if there's no icon:
https://github.com/jhipster/generator-jhipster/blob/e05e776dfd4029ce4b2b27b53d9ddcd430cf64cd/generators/angular/templates/src/main/webapp/app/shared/sort/sort-by.directive.ts.ejs#L55-L57

This PR removes the sortBy / sort icon at direct fields, if the entity is not sortable at all by the given fields (relationship related fields / direct fields) preventing Internal Server error / an inconsistent HTML.

<!--
PR description.
-->

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

<!--
Please also reference the issue number in a commit message to [automatically close the related GitHub issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` or `[ci skip]` to your commit message to skip continuous integration tests
-->

mshima commented on generators/angular/templates/src/main/webapp/app/entities/_entityFolder_/list/_entityFile_.component.html.ejs:
```diff
@@ -73,11 +73,13 @@
-<%_ for (const field of fields.filter(field => !field.hidden)) { _%>
-                    <th scope="col" <%= jhiPrefix %>SortBy="<%= field.fieldName %>">
+<%_ for (const field of fields.filter(field => !field.hidden)) {
+    const fieldIsSortable = !field.transient;
```
I prefer to add as a field property in bootstrap-application generator.

mshima replied:
Rename to fieldSupportsSortBy?

mshima commented on generators/angular/templates/src/main/webapp/app/entities/_entityFolder_/list/_entityFile_.component.html.ejs:
```diff
@@ -88,11 +90,14 @@
 <%_ for (const relationship of relationships.filter(rel => !rel.otherEntityIsEmbedded)) { _%>
   <%_ if (relationship.persistableRelationship && (!relationship.collection || paginationNo)) {
     const fieldName = "." + relationship.otherEntityField;
+    const fieldIsSortable = !relationship.relatedField.transient;
```
I prefer to add as a field property in bootstrap-application generator and maybe rename to `relationshipSupportsSortBy`.

OmarHawk commented:
I've added additional test data and also moved the property into the server generator  - so into the place where it belongs best in my opinion: Sorting itself is being made available by the server component, so I think, it belongs there (this is also the place, where the transient property is set, from which this new property is derived). 

If we would ever have any client side sorting, it could be overriden in the clients prepare field method, otherwise any client would receive this as an info from the server generator to not generate sorting stuff.

mshima commented on generators/angular/templates/src/main/webapp/app/entities/_entityFolder_/list/_entityFile_.component.html.ejs:
```diff
@@ -88,11 +90,14 @@
 <%_ for (const relationship of relationships.filter(rel => !rel.otherEntityIsEmbedded)) { _%>
   <%_ if (relationship.persistableRelationship && (!relationship.collection || paginationNo)) {
     const fieldName = "." + relationship.otherEntityField;
+    const { fieldSupportsSortBy } = relationship.relatedField;
```
Correct me if I am wrong, we only support sorting by relationship id, not relationship label on backend side.
```suggestion
    const { fieldSupportsSortBy: relationshipPkSupportsSortBy } = relationship.primaryKey.fields[0];
```

OmarHawk replied:
No, we actually support sorting by the label, which even makes sense as it would otherwise be really confusing in the UI after applying the sort. 
But that only works if the label field itself is sortable, i.e. not Transient like the mapstructExpression field.

mshima commented on generators/angular/templates/src/main/webapp/app/entities/_entityFolder_/list/_entityFile_.component.html.ejs:
```diff
@@ -88,11 +90,14 @@
 <%_ for (const relationship of relationships.filter(rel => !rel.otherEntityIsEmbedded)) { _%>
   <%_ if (relationship.persistableRelationship && (!relationship.collection || paginationNo)) {
     const fieldName = "." + relationship.otherEntityField;
+    const { fieldSupportsSortBy } = relationship.relatedField;
```
```suggestion
    const { fieldSupportsSortBy: relatedFieldSupportsSortBy } = relationship.relatedField;
```

mshima commented on generators/angular/templates/src/main/webapp/app/entities/_entityFolder_/list/_entityFile_.component.html.ejs:
```diff
@@ -88,11 +90,14 @@
+    const { fieldSupportsSortBy } = relationship.relatedField;
   _%>
-                    <th scope="col" <%= jhiPrefix %>SortBy="<%= relationship.relationshipName + (fieldName) %>">
+                    <th scope="col"<% if (fieldSupportsSortBy) { %> <%= jhiPrefix %>SortBy="<%= relationship.relationshipName + (fieldName) %>"<% } %>>
```
```suggestion
                    <th scope="col"<% if (relatedFieldSupportsSortBy) { %> <%= jhiPrefix %>SortBy="<%= relationship.relationshipName + (fieldName) %>"<% } %>>
```

mshima commented on generators/angular/templates/src/main/webapp/app/entities/_entityFolder_/list/_entityFile_.component.html.ejs:
```diff
@@ -88,11 +90,14 @@
+                    <th scope="col"<% if (fieldSupportsSortBy) { %> <%= jhiPrefix %>SortBy="<%= relationship.relationshipName + (fieldName) %>"<% } %>>
                         <div class="d-flex">
                             <span>__jhiTranslateTag__('<%- i18nKeyPrefix %>.<%- relationship.relationshipName %>')</span>
+    <%_ if (fieldSupportsSortBy) { _%>
```
```suggestion
    <%_ if (relatedFieldSupportsSortBy) { _%>
```

mshima reviewed (approved):
Rename the field for simplicity.
